### PR TITLE
fix: enable kill query before pipeline started.

### PR DESCRIPTION
**Pull request by youngsofun**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

before this pr,  query can only be killed after pipeline is started(` ctx.set_executor()`).

it will be better if long  binding/read_parttiion can be interrupted, which needs another pr.  

Closes  https://github.com/datafuselabs/databend/issues/11302

### CODE REVIEW

File: src/query/service/src/pipelines/executor/pipeline_executor.rs [hunk 1]
```diff
@@ -397,14 +399,17 @@ impl Drop for PipelineExecutor {
...
-        let mut guard = self.on_finished_callback.lock();
-        if let Some(on_finished_callback) = guard.take() {
-            drop(guard);
-            let cause = Some(ErrorCode::Internal(
-                "Pipeline illegal state: not successfully shutdown.",
-            ));
-            if let Err(cause) = catch_unwind(move || on_finished_callback(&cause)).flatten() {
-                warn!("Pipeline executor shutdown failure, {:?}", cause);
+        let inited = *self.inited.lock();
+        if inited {
```
Comment (zhang2014) [inline src/query/service/src/pipelines/executor/pipeline_executor.rs:403]: This will cause to a session leak. we must call 'on_finished'.

Comment (youngsofun) [reply src/query/service/src/pipelines/executor/pipeline_executor.rs:403]: why session leak? drop session in a callback?

Comment (youngsofun) [reply src/query/service/src/pipelines/executor/pipeline_executor.rs:403]: is it ok to add on_finished to  PipelineCompleteExecutor/PipelinePullingExecutor, and call it here

```
impl PipelineCompleteExecutor {
   ...
    pub fn on_finished(&self, cause: Option<ErrorCode>) {
        self.executor.on_finished(&cause).ok();
    }
}

```

**Review by zhang2014 — APPROVED** [review]
.

File: src/query/service/src/interpreters/interpreter.rs [hunk 2]
```diff
@@ -92,17 +97,24 @@ pub trait Interpreter: Sync + Send {
...
+                return Err(err);
+            }
             complete_executor.execute()?;
-            return Ok(Box::pin(DataBlockStream::create(None, vec![])));
+            Ok(Box::pin(DataBlockStream::create(None, vec![])))
+        } else {
+            let pulling_executor = PipelinePullingExecutor::from_pipelines(build_res, settings)?;
+            ctx.set_executor(Arc::downgrade(&pulling_executor.get_inner()));
+            if let Err(err) = ctx.check_aborting() {
+                pulling_executor.on_finished(Some(err.clone()));
```
Comment (zhang2014) [inline src/query/service/src/interpreters/interpreter.rs:110]:
```suggestion
                pulling_executor.finish(Some(err.clone()));
```

**Comment by youngsofun** [comment]
killed by

https://github.com/datafuselabs/databend/blob/main/src/query/service/src/api/rpc/exchange/statistics_receiver.rs#L93

**Comment by youngsofun** [comment]
@zhang2014 plz review again

**Pull request merged**